### PR TITLE
Expose the scaled bitmap interface.

### PR DESCRIPTION
**Pull request by jmoraleda**
Hello Vadim and Robin,

This pull request is to expose the scaled bitmap interface. In particular so that it is available in wxPython. As you know, this interface enables creating bitmaps that are dpi aware. For a more in-depth discussion on the motivation for the pull request see https://discuss.wxpython.org/t/multiplatform-support-for-high-dpi-displays-in-particular-macos/34779

This pull request in particular enables creating a dpi aware bitmap from an image in python, for example as:

```
image = wx.Image('mytestimage.png')
scale = wx.GetApp().GetTopWindow().GetContentScaleFactor()
bitmap = wx.Bitmap(image, -1, scale)
```
Thank you.

### CODE REVIEW

File: interface/wx/bitmap.h [hunk 1]
```diff
@@ -350,8 +350,13 @@ class wxBitmap : public wxGDIObject
...
             Specifies the depth of the bitmap.
             If this is omitted, the display depth of the screen is used.
+        @param scale
+            Specifies the scale for a scaled bitmap. This corresponds to the size
+            of a device-independent standard pixel relative to that of a
+            device-dependent, actual pixel.
+            If this is omitted, 1.0 is used.
     */
-    wxBitmap(const wxImage& img, int depth = wxBITMAP_SCREEN_DEPTH);
+    wxBitmap(const wxImage& img, int depth = wxBITMAP_SCREEN_DEPTH, double scale=1.0);
```
Comment (imciner2) [inline interface/wx/bitmap.h:359]: This argument was only added in 3.1.0, so it should be marked as since 3.1.0.

File: interface/wx/bitmap.h [hunk 2]
```diff
@@ -762,6 +767,31 @@ class wxBitmap : public wxGDIObject
             Bitmap width in pixels.
     */
     virtual void SetWidth(int width);
+
+    /**
+       Retrieves the size of a device-independent standard pixel relative to that
+       of a device-dependent, actual pixel for a scaled bitmap
+       @since 3.1.5
```
Comment (imciner2) [inline interface/wx/bitmap.h:774]: 3.1.5 doesn't seem correct here. These functions were added to the base bitmap class in 3.0 (albeit, the GTK port didn't get native implementations until 3.1.0, and I can't tell if MSW has a native implementation or if it needs it).

**Comment by vadz** [comment]
Sorry for the delay with replying to this. I realize that this is useful functionality, but I'm still not convinced that we actually provide the right/best API for it. Notably, the ctor from `wxImage` has a very real problem: if the image has an odd height or width, using scale factor of 2 just doesn't work. Similarly, `GetScaledSize()` may actually lose pixels. This seems to indicate that this API is not the right one and I really hoped to come with something better before 3.1.4, but unfortunately this is not going to be the case. We do need to resolve this (or resign to living forever with the current imperfect API) before 3.2.0 but I'd like to avoid documenting these methods for now in case we change them.

**Comment by jmoraleda** [comment]
Hello Vadim,

Thank you. I have two thoughts:

First, I fully concur with the issue about loss of pixels not being acceptable in the public API. Furthermore,  `GetContentScaleFactor()` may not return just return `2.0`. I know of at least one case where it returns `1.25` so the issue is not just with bitmaps with odd sizes. The platform in question is windows, so one could argue this example it is not relevant. But I think at the core, we need to accept that there are relevant platforms and bitmap combinations where the scaled size cannot be represented using `int` without loss of precision. To address this, I propose dropping `GetScaledSize` from the public API since the return value `wxSize` internally uses `int`. The other methods take and return `double`, so they don't suffer from this problem. 

Second, I think exposing the ability to create dpi-aware bitmaps in the python api is important and justifies adding the methods to `interface.h` in their present form. If you feel concerns that the API for dpi-aware bitmaps may change in the future, I would suggest that we tag the methods with `@experimental` to warn users that these API endpoints are not part of the stable API and could change in future versions.

**Comment by jmoraleda** [comment]
Another path that I can suggest is for the bitmap to store a scale factor different from the one passed in the ctor. This scale factor would be computed as the nearest float from the one actually passed that yields integer scaled width and height. In this case all methods would return consistent values, and scaled size would be int, but `getScaleFactor` might return a number that is not the one passed in the ctor, which would be a surprising result and thus not desirable.

I don't have a strong preference between both approaches. 

I think in practice potential users of this API are familiar with this issue and choose bitmaps with original sizes that have several twos and threes in their integer factorization. In that case, both approaches yield the same values for the scale factors in common usage.

**Comment by jmoraleda** [comment]
Hello Vadim, 

Thank you again. I've been thinking more. This posting supersedes my previous one, which I leave up for now to not confuse the discussion. This is my current thinking:

`GetContentScaleFactor()` may return values other than `2.0`. I know of at least one case where it returns `1.25` so the issue is not just with bitmaps with odd sizes. It illustrates that there are platforms and image combinations where the scaled size cannot be represented using `int` without loss of precision. 

I believe internally wxWidgets uses _standard pixels_ (i.e. scaled sizes) when performing layout computations. When using scaled sizes and arbitrary sized images, it is possible to have fractional pixels as we've been discussing but since it is not practical to change all size computations everywhere to double or to change all widgets to use actual pixel sizes instead of scaled pixels, the options that remain in my opinion are:

1. Support arbitrary image sizes in wxBitmap using an integer scaled size obtained by rounding . Then we could achieve consistency by returning the rounded scaled width, height and size. To avoid any inconsistency we could avoid being able to retrieve the scale factor in the public api.

2. Support only image sizes that can be scaled exactly. In this case the ctor would throw an exception when the user tries to do a scaling that cannot be done without loss of precision while maintaining integer scaled sizes.

-------------

Regardless of what direction we go, I think exposing the ability to create dpi-aware bitmaps in the python api is important and urgent justifies adding the methods to `interface.h` in their present form. To address the concern that the API for dpi-aware bitmaps may change in the future, I would suggest that we tag the methods with `@experimental` instead of `@since` to warn users that these API endpoints are not part of the stable API and could change in future versions.

**Comment by vadz** [comment]
Yes, scaling factor can definitely be different from 2.0, values 1.5 and 1.25 are common and others are almost surely possible.

wxWidgets uses integers internally for windows, and this is not likely to be changed. For drawing, `wxDC` is still limited to integers, but `wxGraphicsContext` isn't and it would be acceptable (even if not ideal) to just say that you must use the latter for good high DPI support.

This still doesn't solve `wxBitmap` API problem though, as a bitmap, by definition, has an integer number of columns/rows. And I think the only reasonable solution is to store both the logical and the physical size of the bitmap instead of using the scale factor. The current API doesn't even seem to be appropriate for macOS, from where it originated because `wxBitmap::CreateScaled()` can lose precision when rounding, e.g. consider doing `CreateScaled(3, 3, -1 /* depth */, 1.5)`, which will end up doing `CGBitmapContextCreate(4, 4, ...); CGContextScaleCTM(1.5, 1.5)` and I am not at all sure that this is what we want to do here. I believe it would be more useful to fix the size of the window that we'll draw this bitmap in and the actual size of the bitmap and then use the scaling factor equal to their ratio with the window size scaled by DPI. E.g. if we want to draw a bitmap in a window of size `3*3` at DPI 144 (i.e. DPI scaling factor of 1.5), we should somehow decide whether we want to use bitmap of size 4 or 5 and then use either 1.33 or 1.66 effective scaling. @csomor What do you think?


Regarding using `@experimental`, I'm sorry, because I realize that this is not what you want to hear, but I don't think it's a good solution.

**Comment by jmoraleda** [comment]
Unfortunately, I think it is not theoretically possible when using `int` scaled sizes to define a single `double` attribute _scale_ for an arbitrary bitmap that satisfies all the properties one would want it to, even if one does not require the _scale_ attribute to be equal to the _scale_ argument passed in the `ctor`. 

The problem arises with non-square bitmaps. For example, if one has an image that is `8x7` and the _hint_ scale passed to the ctor is `2.0`, after rounding the scaled size would be `4x3` but the real scale would be `2.0` when measured horizontally but `2.333` when measured vertically and something in between when measuring the ratio of the diagonals.

Theoretically, in the absence of other constraints, if we wanted (1) to preserve integer sizes and (2) to support arbitrary image sizes with arbitrary scales, I would recommend dropping the scale attribute of a bitmap. In this approach scale would be used only as a hint in the constructors to pick a scaled width and height so that the ratios between the original width and height and their scaled counterparts are _close_ to  the hint. After that the API would enforce using only width, height and size, both raw and scaled, all `int`, all perfectly consistent and well defined. 

Regrettably, I don't think this works in practice, because, while I have no experience with Apple's appkit at all, I just read about nsimage (https://developer.apple.com/documentation/appkit/nsimage), and it is my impression that the concept of scale is intrinsic to their toolkit. In particular it seems that `wxBitmap` needs to communicate with appkit by giving it the raw image and _a scale_.  I assume that internally appkit uses floating point scaled sizes to avoid all these problems.

In summary, I think that if we want to preserve `int` scaled sizes then it is theoretically necessary to accept either some amount of fuzziness and inconsistency or reject _incompatible_ bitmap sizes and scale combinations. Both options are painful, but I think the issue is important and urgent.

**Comment by csomor** [comment]
I absolutely agree with @vadz that the API offers more breadth than it is intended and supported. On macOS or iOS we don't have fractional scales. And the usage is the other way round, you know you have a high dpi image which is rendered at x2 or x3 and then say so - or ask the OS about it, so that the wxBitmap has its logical dimensions correctly set up and the backing store (raw bitmap) fits the content precisely.

As of now wxBitmap is not supporting multiple resolutions/scales, it is only aware of the scale it carries ...

**Comment by csomor** [comment]
all `Create` methods have the backing store, ie the precise bitmap content, in their first parameters, and the 'interpretation' ie the scale as the last, the only exception is the convenience method `CreateScaled` .. which is used in some places in wx, but this is the only one which could be misused as described, and it could be removed and replaced with explicit calculation of the backing store dimensions and the ordinary `Create` method. `CreateScaled` was not meant to be documented / exposed by this PR anyway

**Comment by csomor** [comment]
I never intended this to be a versatile scale-to-anything functionality - we've got wxImage to transform things - but at as means to support high-dpi devices. So for me asserting when scale would lead to a fractional logical size already in the constructor would absolutely be fine.

**Comment by csomor** [comment]
I've added such asserts to https://github.com/wxWidgets/wxWidgets/pull/1980 

if things build correctly people using high-res images on mac or ios can test.

**Comment by jmoraleda** [comment]
Thank you @csomor. I have added your assertions as a standalone commit also to this pull request since I was not sure whether you intended #1980 to be merged or as a sandbox.

**Comment by vadz** [comment]
Sorry @jmoraleda but I really don't think we want to keep these methods, we're going to have something else for 3.1.5 (see wx-dev discussion), so we're not going to merge this one.

**Comment by jmoraleda** [comment]
Thank you @vadz I look forward to the new API. I am happy we are looking into it.
